zebra-state/src/service/write.rs

```diff
@@ -0,0 +1,131 @@
+//! Writing blocks to the finalized and non-finalized states.
+
+use std::sync::{Arc, Mutex};
+
+use zebra_chain::block::{self, Height};
+
+use crate::service::{
+    finalized_state::FinalizedState,
+    queued_blocks::{QueuedFinalized, QueuedNonFinalized},
+    ChainTipBlock, ChainTipSender,
+};
+
+/// Reads blocks from the channels, write them to the `finalized_state`,
+/// and updates the
+///
+/// TODO: pass the non-finalized state and associated update channel to this function
+#[instrument(skip(
+    finalized_block_write_receiver,
+    non_finalized_block_write_receiver,
+    invalid_block_reset_sender,
+    chain_tip_sender
+))]
+pub fn write_blocks_from_channels(
+    mut finalized_block_write_receiver: tokio::sync::mpsc::UnboundedReceiver<QueuedFinalized>,
+    mut non_finalized_block_write_receiver: tokio::sync::mpsc::UnboundedReceiver<
+        QueuedNonFinalized,
+    >,
+    mut finalized_state: FinalizedState,
+    invalid_block_reset_sender: tokio::sync::mpsc::UnboundedSender<block::Hash>,
+    chain_tip_sender: Arc<Mutex<ChainTipSender>>,
+) {
+    // Write all the finalized blocks sent by the state,
+    // until the state closes the finalized block channel's sender.
+    while let Some(ordered_block) = finalized_block_write_receiver.blocking_recv() {
+        // TODO: split these checks into separate functions
+
+        if invalid_block_reset_sender.is_closed() {
+            info!("StateService closed the block reset channel. Is Zebra shutting down?");
+            return;
+        }
+
+        // Discard any children of invalid blocks in the channel
+        let next_valid_height = finalized_state
+            .db
+            .finalized_tip_height()
+            .map(|height| (height + 1).expect("committed heights are valid"))
+            .unwrap_or(Height(0));
+
+        if ordered_block.0.height > next_valid_height {
+            debug!(
+                ?next_valid_height,
+                invalid_height = ?ordered_block.0.height,
+                invalid_hash = ?ordered_block.0.hash,
+                "got a block that was too high. \
+                 Assuming a parent block failed, and dropping this block",
+            );
+
+            // We don't want to send a reset here, because it could overwrite a valid sent hash
+        }
+
+        // Try committing the block
+        match finalized_state.commit_finalized(ordered_block) {
+            Ok(finalized) => {
+                let tip_block = ChainTipBlock::from(finalized);
+
+                // TODO: update the chain tip sender with non-finalized blocks in this function,
+                //       and get rid of the mutex
+                chain_tip_sender
+                    .lock()
+                    .expect("unexpected panic in block commit task or state")
+                    .set_finalized_tip(tip_block);
```

Comment:
I think this `Mutex` changes the timing of the tip update, which causes some RPC tests that depend on that exact timing to fail:
- `zebra_rpc::methods::tests::snapshot::test_rpc_response_data`
- `zebra_rpc::methods::tests::vectors::rpc_getaddresstxids_response`
- `zebra_rpc::methods::tests::vectors::rpc_getbestblockhash`

https://github.com/ZcashFoundation/zebra/actions/runs/3065213251/jobs/4949092612#step:14:5326
https://github.com/ZcashFoundation/zebra/actions/runs/3065213251/jobs/4949092612#step:14:5746
https://github.com/ZcashFoundation/zebra/actions/runs/3065213251/jobs/4949092612#step:14:5877
https://github.com/ZcashFoundation/zebra/actions/runs/3065213265/jobs/4949322117#step:3:4483

The fix is to wait for the chain tip update _before_ sending the RPC requests: 55e03969ba0bbef4f3f2bb7fa3569ee2668e038f

Reply:
This seems to be causing test hangs, I'll see what I can do.

Reply:
I fixed this by adding a timeout in 770f73a76e